File: shared/src/timeline-markdown.ts
/**
 * Lightweight markdown parser for timeline detail panes.
 *
 * Output is a flat array of typed lines that each platform renders natively
 * (SwiftUI / Compose). Pure data — no rendering. Mirrors the Apple parser at
 * apple/AgentDeck/UI/Monitor/TimelineStripView.swift `TimelineMarkdownLine`
 * and the Android port at
 * android/.../ui/timeline/TimelineMarkdown.kt — keep the three in lockstep.
 *
 * Block grammar (line-oriented):
 *   - ``` toggles a code fence; lines inside become `code` (verbatim, no inline parsing)
 *   - empty / whitespace-only line → `blank`
 *   - 1-6 leading hashes + space → `heading` (level 1-6; CommonMark ATX)
 *   - `- ` or `* ` → `bullet`
 *   - `<digits>.` or `<digits>)` followed by space → `numbered`
 *   - `> ` → `quote`
 *   - `|...|` (with optional `|---|` separator) → `table` block
 *   - anything else → `text`
 *
 * Inline grammar (per `parseInlineSpans`, applied at render time to non-code
 * line content + table cells; first-match-wins, no recursion):
 *   - `` `code` `` → `code` span
 *   - `**bold**`   → `bold` span
 *   - `*italic*`   → `italic` span (only when not adjacent to another `*`)
 *   - `[text](href)` → `link` span
 *   - everything else → `plain` span
 */

export type MarkdownLine =
  | { kind: 'blank' }
  | { kind: 'heading'; level: 1 | 2 | 3 | 4 | 5 | 6; content: string }
  | { kind: 'bullet'; content: string }
  | { kind: 'numbered'; marker: string; content: string }
  | { kind: 'quote'; content: string }
  | { kind: 'code'; content: string }
  | { kind: 'text'; content: string }
  | { kind: 'table'; rows: string[][]; hasHeader: boolean };

export type InlineSpan =
  | { kind: 'plain'; text: string }
  | { kind: 'bold'; text: string }
  | { kind: 'italic'; text: string }
  | { kind: 'code'; text: string }
  | { kind: 'link'; text: string; href: string };

export function parseTimelineMarkdown(text: string): MarkdownLine[] {
  if (!text) return [];
  const out: MarkdownLine[] = [];
  let inCodeFence = false;
  const lines = text.split(/\r?\n/);

  let i = 0;
  while (i < lines.length) {
    const rawLine = lines[i];
    const trimmed = rawLine.trim();

    if (trimmed.startsWith('```')) {
      inCodeFence = !inCodeFence;
      i += 1;
      continue;
    }
    if (inCodeFence) {
      out.push({ kind: 'code', content: rawLine });
      i += 1;
      continue;
    }
    if (trimmed.length === 0) {
      out.push({ kind: 'blank' });
      i += 1;
      continue;
    }

    // Table block: current line looks like `|...|`. Optional separator on
    // next line marks the previous row as header. Continue collecting rows
    // while subsequent lines are also table-rows.
    if (isTableRow(trimmed)) {
      const rows: string[][] = [splitCells(trimmed)];
      let hasHeader = false;
      let j = i + 1;
      if (j < lines.length && isTableSeparator(lines[j].trim())) {
        hasHeader = true;
        j += 1;
      }
      while (j < lines.length) {
        const nextTrimmed = lines[j].trim();
        if (!isTableRow(nextTrimmed) || isTableSeparator(nextTrimmed)) break;
        rows.push(splitCells(nextTrimmed));
        j += 1;
      }
      out.push({ kind: 'table', rows, hasHeader });
      i = j;
      continue;
    }

    const heading = parseHeading(trimmed);
    if (heading) {
      out.push({ kind: 'heading', level: heading.level, content: heading.content });
      i += 1;
      continue;
    }

    if (trimmed.startsWith('- ') || trimmed.startsWith('* ')) {
      out.push({ kind: 'bullet', content: trimmed.slice(2) });
      i += 1;
      continue;
    }

    const numbered = parseNumbered(trimmed);
    if (numbered) {
      out.push({ kind: 'numbered', marker: numbered.marker, content: numbered.content });
      i += 1;
      continue;
    }

    if (trimmed.startsWith('> ')) {
      out.push({ kind: 'quote', content: trimmed.slice(2) });
      i += 1;
      continue;
    }

    out.push({ kind: 'text', content: rawLine });
    i += 1;
  }

  return out.length === 0 ? [{ kind: 'text', content: text }] : out;
}

/**
 * Tokenize a single line of inline content into spans. First-match-wins
 * left-to-right walker; never recurses (so e.g. `**bold *with italic***` is
 * just bold + plain — keeps the implementation simple and the platform ports
 * trivial). Always returns at least one span; empty input → empty array.
 */
export function parseInlineSpans(text: string): InlineSpan[] {
  if (!text) return [];
  const out: InlineSpan[] = [];
  let pending = ''; // buffered plain text waiting to flush
  const flushPlain = () => {
    if (pending.length > 0) {
      out.push({ kind: 'plain', text: pending });
      pending = '';
    }
  };

  let i = 0;
  while (i < text.length) {
    const ch = text[i];

    // `code`
    if (ch === '`') {
      const close = text.indexOf('`', i + 1);
      if (close > i) {
        flushPlain();
        out.push({ kind: 'code', text: text.slice(i + 1, close) });
        i = close + 1;
        continue;
      }
    }

    // **bold**
    if (ch === '*' && text[i + 1] === '*') {
      const close = text.indexOf('**', i + 2);
      if (close > i + 1) {
        flushPlain();
        out.push({ kind: 'bold', text: text.slice(i + 2, close) });
        i = close + 2;
        continue;
      }
    }

    // *italic* — single `*`, not surrounded by another `*`
    if (
      ch === '*' &&
      text[i + 1] !== '*' &&
      (i === 0 || text[i - 1] !== '*')
    ) {
      // Find a closing single `*` that isn't doubled.
      let j = i + 1;
      while (j < text.length) {
        if (text[j] === '*' && text[j + 1] !== '*' && text[j - 1] !== '*') break;
        j += 1;
      }
      if (j < text.length && j > i + 1) {
        flushPlain();
        out.push({ kind: 'italic', text: text.slice(i + 1, j) });
        i = j + 1;
        continue;
      }
    }

    // [text](href)
    if (ch === '[') {
      const closeBracket = text.indexOf(']', i + 1);
      if (closeBracket > i && text[closeBracket + 1] === '(') {
        const closeParen = text.indexOf(')', closeBracket + 2);
        if (closeParen > closeBracket + 1) {
          const linkText = text.slice(i + 1, closeBracket);
          const href = text.slice(closeBracket + 2, closeParen);
          flushPlain();
          out.push({ kind: 'link', text: linkText, href });
          i = closeParen + 1;
          continue;
        }
      }
    }

    pending += ch;
    i += 1;
  }
  flushPlain();
  return out.length === 0 ? [{ kind: 'plain', text }] : out;
}

function parseHeading(trimmed: string): { level: 1 | 2 | 3 | 4 | 5 | 6; content: string } | null {
  let level = 0;
  for (const ch of trimmed) {
    if (ch === '#') level += 1;
    else break;
  }
  if (level < 1 || level > 6) return null;
  if (trimmed.charAt(level) !== ' ') return null;
  return { level: level as 1 | 2 | 3 | 4 | 5 | 6, content: trimmed.slice(level + 1) };
}

function parseNumbered(trimmed: string): { marker: string; content: string } | null {
  const m = trimmed.match(/^(\d+)([.)])\s+(.*)$/);
  if (!m) return null;
  return { marker: `${m[1]}${m[2]}`, content: m[3] };
}

// ---- Table helpers ----

const TABLE_ROW_RE = /^\s*\|.+\|\s*$/;
const TABLE_SEPARATOR_RE = /^\s*\|[-:\s|]+\|\s*$/;

function isTableRow(trimmed: string): boolean {
  return TABLE_ROW_RE.test(trimmed) && !isTableSeparator(trimmed);
}

function isTableSeparator(trimmed: string): boolean {
  if (!TABLE_SEPARATOR_RE.test(trimmed)) return false;
  // Must contain at least one dash to be a real separator (not just `| | |`).
  return /-/.test(trimmed);
}

function splitCells(rowLine: string): string[] {
  return rowLine
    .trim()
    .replace(/^\|/, '')
    .replace(/\|$/, '')
    .split('|')
    .map((c) => c.trim());
}

File: android/app/src/main/kotlin/dev/agentdeck/ui/timeline/TimelineMarkdown.kt
package dev.agentdeck.ui.timeline

/**
 * Parsed markdown line for the timeline detail pane.
 *
 * Mirrors the parser in `shared/src/timeline-markdown.ts` and the Apple
 * `TimelineMarkdownLine` enum in `TimelineStripView.swift`. The grammar is
 * line-oriented; inline spans are computed at render time via
 * [parseInlineSpans] (called from [TimelineMarkdownView]).
 */
sealed class TimelineMarkdownLine {
    object Blank : TimelineMarkdownLine()
    data class Heading(val level: Int, val content: String) : TimelineMarkdownLine()
    data class Bullet(val content: String) : TimelineMarkdownLine()
    data class Numbered(val marker: String, val content: String) : TimelineMarkdownLine()
    data class Quote(val content: String) : TimelineMarkdownLine()
    data class Code(val content: String) : TimelineMarkdownLine()
    data class Plain(val content: String) : TimelineMarkdownLine()

    /**
     * Markdown table block. `rows[0]` is the header when [hasHeader] is true.
     * Cells contain raw inline-markdown text; render through [parseInlineSpans].
     */
    data class Table(val rows: List<List<String>>, val hasHeader: Boolean) : TimelineMarkdownLine()
}

/**
 * Inline span produced by [parseInlineSpans]. Mirrors `InlineSpan` in
 * `shared/src/timeline-markdown.ts`. The renderer turns these into a Compose
 * `AnnotatedString`.
 */
sealed class InlineSpan {
    data class Plain(val text: String) : InlineSpan()
    data class Bold(val text: String) : InlineSpan()
    data class Italic(val text: String) : InlineSpan()
    data class Code(val text: String) : InlineSpan()
    data class Link(val text: String, val href: String) : InlineSpan()
}

/**
 * Parse `text` into a flat list of typed lines for native rendering.
 *
 * Block grammar:
 *   - ``` toggles a code fence; lines inside become Code (verbatim, no inline parsing)
 *   - empty / whitespace-only line → Blank
 *   - 1..6 leading hashes followed by a space → Heading
 *   - "- " or "* " → Bullet
 *   - `<digits>.` or `<digits>)` + space → Numbered
 *   - "> " → Quote
 *   - `|...|` (with optional `|---|` separator) → Table block
 *   - anything else → Plain
 */
fun parseTimelineMarkdown(text: String): List<TimelineMarkdownLine> {
    if (text.isEmpty()) return emptyList()
    val out = mutableListOf<TimelineMarkdownLine>()
    var inCodeFence = false
    val lines = text.split('\n').map { it.trimEnd('\r') }

    var i = 0
    while (i < lines.size) {
        val rawLine = lines[i]
        val trimmed = rawLine.trim()

        if (trimmed.startsWith("```")) {
            inCodeFence = !inCodeFence
            i += 1
            continue
        }
        if (inCodeFence) {
            out += TimelineMarkdownLine.Code(rawLine)
            i += 1
            continue
        }
        if (trimmed.isEmpty()) {
            out += TimelineMarkdownLine.Blank
            i += 1
            continue
        }

        // Table block — current line is `|...|` and not a separator.
        if (isTableRow(trimmed)) {
            val rows = mutableListOf(splitCells(trimmed))
            var hasHeader = false
            var j = i + 1
            if (j < lines.size && isTableSeparator(lines[j].trim())) {
                hasHeader = true
                j += 1
            }
            while (j < lines.size) {
                val nextTrimmed = lines[j].trim()
                if (!isTableRow(nextTrimmed)) break
                rows += splitCells(nextTrimmed)
                j += 1
            }
            out += TimelineMarkdownLine.Table(rows = rows, hasHeader = hasHeader)
            i = j
            continue
        }

        val heading = parseHeading(trimmed)
        if (heading != null) {
            out += TimelineMarkdownLine.Heading(heading.first, heading.second)
            i += 1
            continue
        }

        if (trimmed.startsWith("- ") || trimmed.startsWith("* ")) {
            out += TimelineMarkdownLine.Bullet(trimmed.substring(2))
            i += 1
            continue
        }

        val numbered = NUMBERED_RE.matchEntire(trimmed)
        if (numbered != null) {
            out += TimelineMarkdownLine.Numbered(
                marker = numbered.groupValues[1] + numbered.groupValues[2],
                content = numbered.groupValues[3],
            )
            i += 1
            continue
        }

        if (trimmed.startsWith("> ")) {
            out += TimelineMarkdownLine.Quote(trimmed.substring(2))
            i += 1
            continue
        }

        out += TimelineMarkdownLine.Plain(rawLine)
        i += 1
    }

    return if (out.isEmpty()) listOf(TimelineMarkdownLine.Plain(text)) else out
}

/**
 * Tokenize a single line of inline content into spans. First-match-wins
 * left-to-right walker; never recurses. Mirrors the TS [parseInlineSpans] in
 * `shared/src/timeline-markdown.ts`.
 */
fun parseInlineSpans(text: String): List<InlineSpan> {
    if (text.isEmpty()) return emptyList()
    val out = mutableListOf<InlineSpan>()
    val pending = StringBuilder()
    fun flushPlain() {
        if (pending.isNotEmpty()) {
            out += InlineSpan.Plain(pending.toString())
            pending.clear()
        }
    }
    var i = 0
    while (i < text.length) {
        val ch = text[i]

        // `code`
        if (ch == '`') {
            val close = text.indexOf('`', i + 1)
            if (close > i) {
                flushPlain()
                out += InlineSpan.Code(text.substring(i + 1, close))
                i = close + 1
                continue
            }
        }

        // **bold**
        if (ch == '*' && i + 1 < text.length && text[i + 1] == '*') {
            val close = text.indexOf("**", i + 2)
            if (close > i + 1) {
                flushPlain()
                out += InlineSpan.Bold(text.substring(i + 2, close))
                i = close + 2
                continue
            }
        }

        // *italic* — single `*`, not adjacent to another `*`
        if (ch == '*' &&
            (i + 1 >= text.length || text[i + 1] != '*') &&
            (i == 0 || text[i - 1] != '*')) {
            var j = i + 1
            while (j < text.length) {
                if (text[j] == '*') {
                    val prev = if (j > 0) text[j - 1] else ' '
                    val next = if (j + 1 < text.length) text[j + 1] else ' '
                    if (prev != '*' && next != '*') break
                }
                j += 1
            }
            if (j < text.length && j > i + 1) {
                flushPlain()
                out += InlineSpan.Italic(text.substring(i + 1, j))
                i = j + 1
                continue
            }
        }

        // [text](href)
        if (ch == '[') {
            val closeBracket = text.indexOf(']', i + 1)
            if (closeBracket > i &&
                closeBracket + 1 < text.length &&
                text[closeBracket + 1] == '(') {
                val closeParen = text.indexOf(')', closeBracket + 2)
                if (closeParen > closeBracket + 1) {
                    flushPlain()
                    out += InlineSpan.Link(
                        text = text.substring(i + 1, closeBracket),
                        href = text.substring(closeBracket + 2, closeParen),
                    )
                    i = closeParen + 1
                    continue
                }
            }
        }

        pending.append(ch)
        i += 1
    }
    flushPlain()
    return if (out.isEmpty()) listOf(InlineSpan.Plain(text)) else out
}

private val NUMBERED_RE = Regex("""^(\d+)([.)])\s+(.*)$""")

private fun parseHeading(trimmed: String): Pair<Int, String>? {
    var level = 0
    for (ch in trimmed) {
        if (ch == '#') level += 1 else break
    }
    if (level !in 1..6) return null
    if (level >= trimmed.length || trimmed[level] != ' ') return null
    return level to trimmed.substring(level + 1)
}

// ---- Table helpers ----

private val TABLE_ROW_RE = Regex("""^\s*\|.+\|\s*$""")
private val TABLE_SEPARATOR_INNER = setOf('-', ':', ' ', '|', '\t')

private fun isTableRow(trimmed: String): Boolean {
    if (!TABLE_ROW_RE.matches(trimmed)) return false
    return !isTableSeparator(trimmed)
}

private fun isTableSeparator(trimmed: String): Boolean {
    if (!trimmed.startsWith("|") || !trimmed.endsWith("|") || trimmed.length < 2) return false
    val inner = trimmed.substring(1, trimmed.length - 1)
    if (!inner.contains('-')) return false
    return inner.all { it in TABLE_SEPARATOR_INNER }
}

private fun splitCells(rowLine: String): List<String> {
    val inner = rowLine.trim().removePrefix("|").removeSuffix("|")
    return inner.split('|').map { it.trim() }
}
